akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -491,24 +490,42 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
         None // don't change state
       }
     }
+    def onTrailingHeaders(headers: ParsedHeadersFrame): Unit = {
+      trailingHeaders = wrapTrailingHeaders(headers)
+      if (headers.endStream) {
+        onDataFrame(DataFrame(headers.streamId, endStream = true, ByteString.empty)) // simulate end stream by empty dataframe
+      } else pushGOAWAY(Http2Protocol.ErrorCode.PROTOCOL_ERROR, "Got unexpected mid-stream HEADERS frame")
+    }
     def onRstStreamFrame(rst: RstStreamFrame): Unit = {
       outlet.fail(new PeerClosedStreamException(rst.streamId, rst.errorCode))
       buffer = ByteString.empty
+      trailingHeaders = None
       wasClosed = true
     }
 
     private def dispatchNextChunk(): Unit = {
       if (buffer.nonEmpty && outlet.isAvailable) {
         val dataSize = buffer.size min settings.requestEntityChunkSize
-        outlet.push(buffer.take(dataSize))
+        outlet.push(wrapData(buffer.take(dataSize)))
         buffer = buffer.drop(dataSize)
 
         totalBufferedData -= dataSize
 
         debug(s"Dispatched chunk of $dataSize for stream [$streamId], remaining window space now $outstandingStreamWindow, buffered: ${buffer.size}")
         updateWindows()
       }
-      if (buffer.isEmpty && wasClosed) outlet.complete()
+      if (buffer.isEmpty && wasClosed) {
+        trailingHeaders match {
+          case Some(trailer) =>
+            if (outlet.isAvailable) {
+              outlet.push(trailer)
+              trailingHeaders = None
+              outlet.complete()
+            }
```

Comment:
Ok, so in the scenario where it was not available, we leave here, and on the next pull we will end up here again.